### ISSUE TITLE
HTTP 500 - Memory Exhausted when downloading recordings

### ISSUE THREAD

**Issue by keybrdist** [AzuraCast/AzuraCast#6880]
### Installation Method

Docker Installation

### AzuraCast Release Channel

Rolling Release Channel

### Current AzuraCast Version

Rolling Release #4d81179 (2023-12-28 21:10)

### What happened?

The HTTP Response is a generic HTTP 500 Internal Server Error.

System log:
```log
[2024-01-21T06:52:01.055567+00:00] AzuraCast.ALERT: Fatal Error (E_ERROR): Allowed memory size of 1047527424 bytes exhausted (tried to allocate 389364272 bytes) {"code":1,"message":"Allowed memory size of 1047527424 bytes exhausted (tried to allocate 389364272 bytes)","file":"/var/azuracast/www/vendor/spiral/goridge/src/Frame.php","line":119,"trace":null} []
[2024-01-21T06:52:01.055859+00:00] AzuraCast.ALERT: Fatal Error (E_ERROR): Allowed memory size of 1047527424 bytes exhausted (tried to allocate 389364272 bytes) {"code":1,"message":"Allowed memory size of 1047527424 bytes exhausted (tried to allocate 389364272 bytes)","file":"/var/azuracast/www/vendor/spiral/goridge/src/Frame.php","line":119,"trace":null} []
[2024-01-21T06:52:01.056013+00:00] AzuraCast.ALERT: Fatal Error (E_ERROR): Allowed memory size of 1047527424 bytes exhausted (tried to allocate 389364272 bytes) {"code":1,"message":"Allowed memory size of 1047527424 bytes exhausted (tried to allocate 389364272 bytes)","file":"/var/azuracast/www/vendor/spiral/goridge/src/Frame.php","line":119,"trace":null} []
```

I suspect some part of this process is trying to load the audio file into memory rather than doing streamed response.

- Our recording sizes are typically 300-700MB+
- Storage type: AWS S3 Bucket
- Example: endpoint : `/api/station/1/streamer/12/broadcast/173/download`
^^ (taken from the recording download link)

**Comment by Moonbase59** [comment]
Just checked quickly using Rolling Release #[87e7802](https://github.com/AzuraCast/AzuraCast/commit/87e7802d01b3dd366b8ef4b668e3dcd54e0ba6bf) (2024-01-21 3:08), and it works here. I don’t have large recordings currently, and rather generous PHP memory settings.

In your `/var/azuracast/azuracast.env`, you might want to check the `PHP_MEMORY_LIMIT`:

```
PHP_MEMORY_LIMIT=2G
```

As you see, I have mine at a rather generous 2 GB (for other reasons). Maybe that helps in your case, too. You need to down/up your Docker if changing this parameter:

```bash
cd /var/azuracast
docker-compose down
docker-compose up -d
```

(You might need to use `docker compose` instead of `docker-compose`, depending on your Docker setup. Also, prefixing it with `sudo ` might be needed.

And of course your VM must have enough memory.

**Comment by keybrdist** [comment]
I appreciate the input. 

I did already raise my memory limits before I posted this, but to expand -- raising the memory further will help slightly, but I'm afraid it will not solve the underlying issue, because you can still run out of memory; for example if more than one administrator is logged in and tries to download files simultaneously.  

This does not happen if I serve the static files directly from the nginx service rather than through the PHP app; which tells me PHP is loading these files into memory for some reason.

So,  I'm hoping the AzuraCast team can do some test to make double sure that PHP is not reading files into memory; and that the File Adapters are properly working when using the ->stream() method.  I'm a developer, so I took a brief look at the code and it appears the FileAdapters SHOULD already be doing this, BUT the fact we're still seeing "out of memory" exceptions, it gives me the strong impression that the file stream method is not actually working.  

The PHP application should at the very least be doing some byte-chunking so that parts of the file are chunked out instead of loading the full file into memory.

Btw, I think this is might be affecting the `/play` media endpoints as well, because I was trying to listen back to a 800MB recording a few weeks ago and it completely locked up the entire application, took the station offline (until I restarted the container). I haven't dared tried to do that again because this instance of Azura is running on a production environment serving an established radio station.  

Anyway. I hope this helps!

**Comment by keybrdist** [comment]
One last thought; since these files are served over S3; perhaps you can just use the temporaryUrl() method provided by  FlySystem and use that as the /download and /play endpoints. This should fix the issue completely.

**Comment by Moonbase59** [comment]
Hm. Just to be sure, I just looked up the largest song I have (Klaus Schulze - Opera Trance, 79:20 mins, 558.3 MB FLAC) and played it through the "Media" page. Downloading it from there worked too, took only a few seconds. No irregularities found watching the Admin Status (CPU load, memory, etc.)

So either something was changed between our RR versions, or it differs between recordings and songs. My setup is a Ubuntu 22.04.3 LTS VM, 4 CPUs, 4 GB RAM, and I currently serve 4 stations accessing the same music collection over an NFS mount, having ~170,000 tracks and ~60 playlists/station.

**Comment by BusterNeece** [comment]
@keybrdist Yeah...if you're serving from S3 then the dynamics are totally different than serving from a local filesystem. If you're on a local filesystem, we expressly _don't_ go through PHP to stream the file and instead use nginx's `X-Accel-Redirect` setting to tell nginx to directly serve the file off disk itself.

For remote filesystems, right now we `readStream()`, but for the entire stream to be served back via Roadrunner, it has to load the whole response into memory, which results in the whole file being stored in PHP momentarily.

For S3, the temporaryURL thing might be an effective solution to the problem, but I want to make sure we implement this robustly in a way that works across all of our file serving code.

**Comment by Moonbase59** [comment]
So a remote NFS mount would be considered "local"? That’s good to know, thanks for the explanation.

**Comment by BusterNeece** [comment]
@Moonbase59 As far as our app is concerned yep.

**Comment by keybrdist** [comment]
Thanks to both of you for the help!

@BusterNeece, In the near future, I'll likely fork AC and do some PR's for this and other matters. Happy to contribute to this project if that's cool with you.  

Meanwhile, can you let me know what other considerations I should keep in mind if I implement the temporaryUrl strategy. I want to be sure that it handles any concerns you might have re: robustness of the implementation.

**Comment by BusterNeece** [comment]
@keybrdist I'm always happy for any PRs! In this case, it might not be necessary, though...I'm _reasonably_ certain that if I modify the `App\Http\Response` `streamFilesystemFile` (or whatever it's called) to check for an S3 object and return a 302 to the temporary URL, that would work fine...though that would mean for every single request coming in, one of those temporary URLs would be created. This would be 100% fine for the admin side of things (downloading recordings, etc.) and 0% fine for public-facing album art, where we really really should use nginx's local caching instead. May need to make it a boolean flag.

**Comment by keybrdist** [comment]
Just wondering if you will be patching this issue soonish?

**Comment by BusterNeece** [comment]
After a similar report on a different part of the application, I looked into our configuration on production servers and discovered that our Roadrunner configuration was set to never send "chunked responses" for large files back, which meant that it loaded the entire file contents into memory if it couldn't serve them directly off the filesystem (as it would with local filesystems). I've tweaked our configuration on this, and I can confirm that this now results in significantly fewer memory issues when serving larger files. I've also updated some related code on our backend to use the "accelerated" response serving in more cases. Hopefully this will mean smoother experiences for folks using S3 buckets and other remote filesystems.
